### PR TITLE
Make pausing on window focus loss instantaneous

### LINKED ISSUE #11624: Unpausing the game does not require you to put your cursor where it was if the game was paused via tabbing out
As the title says, when the game pauses due to tabbing out (alt tabbing or another program stealing focus) the game does not require you to go through the process of putting your cursor back where it was, like you normally would. I think part of the problem is that it does not pause immediately as I was actually able to tab out and back in once without it pausing.

Video [here](https://youtu.be/jZPvDh7RDfA) of me demonstrating. The black screen is me tabbing out, it takes a second for OBS to re-focus on osu! after tabbing back in.

osu! lazer 2021.129.0

I don't really think they will be helpful, but here are my logs:

[network.log](https://github.com/ppy/osu/files/5892300/network.log)
[performance.log](https://github.com/ppy/osu/files/5892301/performance.log)
[runtime.log](https://github.com/ppy/osu/files/5892302/runtime.log)
[updater.log](https://github.com/ppy/osu/files/5892297/updater.log)
[database.log](https://github.com/ppy/osu/files/5892298/database.log)

### PR DESCRIPTION
**Pull request by frenzibyte**
Closes https://github.com/ppy/osu/issues/11624 as per https://github.com/ppy/osu/issues/11624#issuecomment-770118396

I've moved the "pause on focus lost" logic outside of the button and into `Player` while at it, given that the button is no longer relevant in pausing for this case.

I've also corrected `IsBreakTime` to become `true` by default, and calculate its value directly in the property to avoid incorrectly setting it to `false` there.

Cannot add a test case (unfortunately) for the above issue as there's no window in a headless host.

### CODE REVIEW

File: osu.Game/Screens/Play/BreakTracker.cs [hunk 1]
```diff
@@ -45,8 +45,12 @@ public BreakTracker(double gameplayStartTime = 0, ScoreProcessor scoreProcessor
         protected override void Update()
         {
             base.Update();
+            updateBreakTime();
+        }
 
-            var time = Clock.CurrentTime;
+        private void updateBreakTime()
+        {
+            var time = Clock?.CurrentTime ?? 0;
```
Comment (bdach) [inline osu.Game/Screens/Play/BreakTracker.cs:53]: I'm not sure this should be doing anything at all if there's no clock available.

(Also a drive-by change with no clear justification in the description.)

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -170,7 +179,7 @@ protected virtual void PrepareReplay()
         }
 
         [BackgroundDependencyLoader(true)]
-        private void load(AudioManager audio, OsuConfigManager config, OsuGame game)
+        private void load(AudioManager audio, OsuConfigManager config, OsuGame game, OsuGameBase gameBase)
```
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:182]: Injected parameter is not used.

File: osu.Game/Screens/Play/Player.cs [hunk 3]
```diff
@@ -154,6 +159,10 @@ protected override void LoadComplete()
             // replays should never be recorded or played back when autoplay is enabled
             if (!Mods.Value.Any(m => m is ModAutoplay))
                 PrepareReplay();
+
+            // needs to be bound here as the last binding, otherwise cases like starting a replay while not focused causes player to exit, if activity is bound before checks.
```
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:163]: Can you explain? I would like to think the `HasReplayLoaded` check in the callback should suffice.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/BreakTracker.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/BreakTracker.cs:53]: Yeah fair point

File: osu.Game/Screens/Play/Player.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:163]: if game activity is bound before the guards are and game is inactive, then calling the method before all guards are set to correct values may result in incorrect actions performed.

but actually now that you said it, I think it could be removed as I've corrected `IsBreakTime` to stay `true` until it's cleared on first `Update()`.

Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:163]: I don't think `IsBreakTime` should be `true` by default.

And you haven't actually answered my question, so I'll answer it myself: the `PrepareReplay()` call *sets* `HasReplayLoaded`. I'd maybe reword the comment to mention that more directly.

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:163]: ah, I haven't actually thought about the `PrepareReplay()` call above, just put this in a place where it should be the last thing to be called, will reword then.

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:163]: hmm, since I've changed `performUserRequestedExit` to calling `Pause()` directly, the comment looks to be completely redundant, I'm a bit concerned about whether to remove it as that sounds like it would hide a potential issue.. what do you think?

Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:163]: If the order of operations no longer matters then I'd just remove the comment.

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -420,10 +426,14 @@ private void updateSampleDisabledState()
             samplePlaybackDisabled.Value = DrawableRuleset.FrameStableClock.IsCatchingUp.Value || GameplayClockContainer.GameplayClock.IsPaused.Value;
         }
 
-        private void updatePauseOnFocusLostState() =>
-            HUDOverlay.HoldToQuit.PauseOnFocusLost = PauseOnFocusLost
-                                                     && !DrawableRuleset.HasReplayLoaded.Value
-                                                     && !breakTracker.IsBreakTime.Value;
+        private void updatePauseOnFocusLostState()
+        {
+            if (!PauseOnFocusLost || breakTracker.IsBreakTime.Value)
```
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:431]: Why is the "replay loaded" check gone now? I don't recall requesting removing it.

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:431]: It's not something to care about anymore given that this calls `Pause()` directly instead of `performUserRequestedExit` previously (where if a replay is loaded it'll exit from player immediately)

Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:431]: Half an answer again... `canPause` inside of `Pause()` checks `HasReplayLoaded` anyway is the actual complete answer.

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:431]: image
I've put it in the description thought you would read it, but sure will try answer completely next time.

**Review by bdach — DISMISSED** [review]

**Comment by peppy** [comment]
Haven't checked the code, but this doesn't seem to work correctly if you make the window inactive too soon after the last pause.

**Comment by peppy** [comment]
I guess this isn't a regression and also happens on master if you set your hold-to-pause time way down.

**Review by peppy — APPROVED** [review]
